File: auto_verbs/sites-not-suitable-SPA/parkrun_org/parkrun_search.py
import os
import sys
import shutil
import time
from dataclasses import dataclass, field
from typing import List

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
from cdp_utils import get_free_port, get_temp_profile_dir, launch_chrome, wait_for_cdp_ws
from playwright_debugger import checkpoint


@dataclass(frozen=True)
class ParkrunSearchRequest:
    search_query: str = "London"
    max_results: int = 5


@dataclass
class ParkrunSearchItem:
    event_name: str = ""
    location: str = ""
    country: str = ""
    distance: str = ""
    day_of_week: str = ""
    avg_runners: str = ""


@dataclass
class ParkrunSearchResult:
    items: List[ParkrunSearchItem] = field(default_factory=list)


def parkrun_search(page, request: ParkrunSearchRequest) -> ParkrunSearchResult:
    url = f"https://www.parkrun.com/countries/"
    page.goto(url, wait_until="domcontentloaded")
    page.wait_for_timeout(5000)

    items = page.evaluate("""() => {
        const results = [];
        const seen = new Set();
        // Get links to parkrun events
        const links = document.querySelectorAll('a[href]');
        for (const a of links) {
            const href = a.getAttribute('href') || '';
            // Match parkrun event pages like /eventname/
            if (!/parkrun\\.(us|com|org|co\\.uk)\\/[a-z]/.test(href)) continue;
            if (/\\/(events|about|register|results|news|faq|volunteer|freedom)/.test(href)) continue;
            if (seen.has(href)) continue;
            seen.add(href);
            const text = a.textContent.trim();
            if (!text || text.length < 3 || text.length > 100) continue;
            results.push({
                event_name: text,
                location: '',
                country: 'United States',
                distance: '5k',
                day_of_week: 'Saturday',
                avg_runners: ''
            });
        }
        return results;
    }""")

    result = ParkrunSearchResult()
    for item in items[: request.max_results]:
        result.items.append(
            ParkrunSearchItem(
                event_name=item.get("event_name", ""),
                location=item.get("location", "") or request.search_query,
                country=item.get("country", ""),
                distance=item.get("distance", "5k"),
                day_of_week=item.get("day_of_week", "Saturday"),
                avg_runners=item.get("avg_runners", ""),
            )
        )

    checkpoint("parkrun_search result")
    return result


def test_func():
    port = get_free_port()
    profile_dir = get_temp_profile_dir()
    chrome_process = launch_chrome(profile_dir, port)
    ws_url = wait_for_cdp_ws(port)

    from playwright.sync_api import sync_playwright

    pw = sync_playwright().start()
    browser = pw.chromium.connect_over_cdp(ws_url)
    context = browser.contexts[0]
    page = context.pages[0] if context.pages else context.new_page()

    try:
        request = ParkrunSearchRequest(search_query="London", max_results=5)
        result = parkrun_search(page, request)
        print(f"Found {len(result.items)} parkrun events")
        for i, item in enumerate(result.items):
            print(f"  {i+1}. {item.event_name}")
            print(f"     Location: {item.location} | Country: {item.country}")
            print(f"     Distance: {item.distance} | Day: {item.day_of_week} | Avg Runners: {item.avg_runners}")
    finally:
        browser.close()
        pw.stop()
        chrome_process.terminate()
        shutil.rmtree(profile_dir, ignore_errors=True)


if __name__ == "__main__":
    from playwright_debugger import run_with_debugger

    run_with_debugger(test_func)

File: auto_verbs/sites-not-suitable-SPA/parkrun_org/parkrun_search.js
const { Stagehand } = require("@browserbasehq/stagehand");
const { PlaywrightRecorder, setupLLMClient } = require("../../stagehand-utils");
const fs = require("fs");
const path = require("path");

const CFG = {
  search_query: "London",
  max_results: 5,
};

async function genPython(cfg, recorder) {
  const url = `https://www.parkrun.com/events/`;

  recorder.addStep(`page.goto("${url}")`);
  recorder.addStep(`page.wait_for_load_state("networkidle")`);
  recorder.addStep(`import time; time.sleep(2)`);

  // Try to search for the location
  recorder.addStep(`search_input = page.query_selector('input[type="search"], input[type="text"], #search, [class*="search"]')`);
  recorder.addStep(`if search_input:
    search_input.fill("${cfg.search_query}")
    search_input.press("Enter")
    page.wait_for_load_state("networkidle")
    time.sleep(2)`);

  recorder.addStep(`items = page.evaluate("""() => {
    const results = [];
    const rows = document.querySelectorAll('tr, [class*="event"], [class*="result"], li[class*="park"]');
    for (let i = 0; i < Math.min(rows.length, ${cfg.max_results}); i++) {
      const el = rows[i];
      if (!el.innerText.trim()) continue;
      const cells = el.querySelectorAll('td');
      const nameEl = el.querySelector('a, [class*="name"], [class*="title"]');
      const locationEl = cells.length > 1 ? cells[1] : el.querySelector('[class*="location"]');
      const countryEl = cells.length > 2 ? cells[2] : el.querySelector('[class*="country"]');
      results.push({
        event_name: nameEl ? nameEl.innerText.trim() : (cells.length > 0 ? cells[0].innerText.trim() : ''),
        location: locationEl ? locationEl.innerText.trim() : '',
        country: countryEl ? countryEl.innerText.trim() : '',
        distance: '5k',
        day_of_week: 'Saturday',
        avg_runners: ''
      });
    }
    return results.filter(r => r.event_name);
  }""")`);

  recorder.addStep(`result = {"items": items}`);
}

(async () => {
  const stagehand = new Stagehand({
    env: "LOCAL",
    verbose: 2,
    model: "openai/gpt-4.1-mini",
  });
  await stagehand.init();
  const page = stagehand.context.pages()[0];

  const recorder = new PlaywrightRecorder();
  await genPython(CFG, recorder);

  const outputPath = path.join(__dirname, "parkrun_search.py");
  console.log("Generated steps:", recorder.getSteps());

  await stagehand.close();
})();
